### PR TITLE
detect: allow rule which need both directions to match

### PR DESCRIPTION
**Pull request by catenacyber**
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/5665

Describe changes:
- allows bidirectional signature matching !

SV_BRANCH=https://github.com/OISF/suricata-verify/pull/1922

https://github.com/OISF/suricata/pull/11528 with better doc bullet list

### CODE REVIEW

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 21630

**Review by jufajardini — CHANGES_REQUESTED** [review]
Left minor comments for docs.

Managed to review up until `detect-engine.c` (around https://github.com/OISF/suricata/pull/11530/files#diff-f0400524fccdb926fe9697ed027b4210459222917a866616235d162b880370f2R1449). So far, seems good!

Following up on a previous PR comment about name suggestions, what about transaction rule? (or something that brought attention to this aspect) Asking also because the code itself refers to `<>` rules as bidir rules, so if we could avoid possible misunderstandings there, that'd be great.

One question that emerged while reviewing was: what happens if a rule tries to use negated or absent content?

**Comment by inashivb** [comment]
> Following up on a previous PR comment about name suggestions, what about transaction rule? (or something that brought attention to this aspect) Asking also because the code itself refers to `<>` rules as bidir rules, so if we could avoid possible misunderstandings there, that'd be great.

+1

**Comment by catenacyber** [comment]
> One question that emerged while reviewing was: what happens if a rule tries to use negated or absent content?

This should work as unidirectional rules that use negated content...
Do you have an example ?

File: doc/userguide/rules/intro.rst [hunk 1]
```diff
@@ -248,10 +251,53 @@ Now, let's say we have a rule with the following header::
...
+Bidirectional rules have some limitations :
+
+* They are only meant to work on transactions with first a request to the server,
+  and then a response to the client, and not the other way around (not tested).
+* They cannot have ``fast_pattern`` or ``prefilter`` the direction to client
+  if they also have a streaming buffer on the direction to server, see example below.
+* They will refuse to load if a single directional rule is enough.
+
+This rule cannot have the ``fast_pattern`` to client, as ``file.data`` is a streaming buffer.
+.. container:: example-rule
```
[the remark this answers is not in the document]
Comment (catenacyber) [reply doc/userguide/rules/intro.rst:297]: ok fixing

File: src/detect-flow.c [hunk 2]
```diff
@@ -79,6 +79,48 @@ void DetectFlowRegister (void)
...
+
+/**
+ * \brief Registration function for flow: keyword
+ */
+void DetectBidirRegister(void)
+{
+    sigmatch_table[DETECT_BIDIR_TOCLIENT].name = "bidir.toclient";
+    sigmatch_table[DETECT_BIDIR_TOCLIENT].desc =
+            "match next keywords only toclient side for bidirectional rules";
+    sigmatch_table[DETECT_BIDIR_TOCLIENT].url = "/rules/flow-keywords.html#TODO";
```
[the remark this answers is not in the document]
Comment (catenacyber) [reply src/detect-flow.c:112]: Thanks, fixing this

File: src/detect-engine-mpm.c [hunk 3]
```diff
@@ -1066,6 +1066,24 @@ static SigMatch *GetMpmForList(const Signature *s, SigMatch *list, SigMatch *mpm
 
 int g_skip_prefilter = 0;
 
+bool DetectBufferToClient(const DetectEngineCtx *de_ctx, int buf_id, AppProto alproto)
+{
+    bool r = false;
+    const DetectEngineAppInspectionEngine *app = de_ctx->app_inspect_engines;
+    for (; app != NULL; app = app->next) {
+        if (app->sm_list == buf_id && (AppProtoEquals(alproto, app->alproto) || alproto == 0)) {
+            if (app->dir == 1) {
```
[the remark this answers is not in the document]
Comment (catenacyber) [reply src/detect-engine-mpm.c:1075]: You can create an optimization ticket (if there is not one already), and it should be a boolean like `if (app->dir_toclient)`

**Comment by catenacyber** [comment]
Thanks Juliana, continued in https://github.com/OISF/suricata/pull/11561

**Comment by jufajardini** [comment]
> > One question that emerged while reviewing was: what happens if a rule tries to use negated or absent content?
> 
> This should work as unidirectional rules that use negated content... Do you have an example ?

I'll try to come up with one (and see if I manage to test it using the corresponding SV branch for the most recent version of this work :) )

File: src/detect-engine-mpm.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (jufajardini) [reply src/detect-engine-mpm.c:1075]: Ticket created - https://redmine.openinfosecfoundation.org/issues/7186

**Pull request closed without merging**